### ISSUE TITLE
TCH004: local imports shadowing TYPE_CHECKING imports lead to false positives

### ISSUE THREAD

**Issue by twoertwein** [astral-sh/ruff#6928]
A simplified version of what happened in https://github.com/pandas-dev/pandas/pull/54786

```py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from bar import Foo # will be moved out by ruff

def test() -> Foo:  # the Foo from TYPE_CHECKING
    # avoid cyclic imports (yes, this is bad design)
    from bar import Foo

    return Foo() # this is the local Foo, the TYPE_CHECKING Foo is not "used"
```

edit: The issue was something else in https://github.com/pandas-dev/pandas/pull/54786 it was actually used at runtime inside a cast statement.

**Comment by zanieb** [comment]
Thanks for the report! I think this is a fair pattern to special case.

**Comment by twoertwein** [comment]
Ruff actually already supports this (and ruff was moving it out because I actually used it in a cast statement).

Sorry for the noise!
